### PR TITLE
Extract Method Can Call ConfigureAwait(False) if extracted method does.

### PR DESCRIPTION
**Pull request by YairHalberstadt**
See #38529

If anywhere in the extracted method we call ConfigureAwait false, we call ConfigureAwait(false) in the invocation of the extracted method.

This preserves semantics around configure await, since if ConfigureAwait(false) is called anywhere in a method, the rest of the method will not continue on the original captured context.

I also fixed a bug at the same time, namely now Extract Method does not make the extracted method async, if it only used await in a local function.

### CODE REVIEW

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 1]
```diff
@@ -578,6 +578,29 @@ protected override ExpressionSyntax CreateCallSignature()
...
 
+                if (this.CSharpSelectionResult.ShouldCallConfigureAwaitFalse())
+                {
+                    if (this.AnalyzerResult.ReturnType.GetMembers().Any(x => x is IMethodSymbol
+                    {
+                        Name: "ConfigureAwait",
+                        Parameters:
+                        {
+                            Length: 1
+                        } parameters
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: let's play around with this a bit.  I think it would be good to have a balance of horizotnal and vertical matching.  How about:

```c#
 x is IMethodSymbol
{
  Name: "ConfigureAwait",
  Parameters: {  Length: 1 } parameters
} &&
parameters[0].Type is { SpecialType: SpecialType.System_Boolean }
```

Basically, be willing to inline simple property patterns to not feel like a simple concept (a 1-length parameter array) needs so many lines of code to express.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 2]
```diff
@@ -578,6 +578,29 @@ protected override ExpressionSyntax CreateCallSignature()
                     return invocation;
                 }
 
+                if (this.CSharpSelectionResult.ShouldCallConfigureAwaitFalse())
+                {
+                    if (this.AnalyzerResult.ReturnType.GetMembers().Any(x => x is IMethodSymbol
+                    {
+                        Name: "ConfigureAwait",
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: `nameof(ConfigureAwait)`

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 3]
```diff
@@ -578,6 +578,29 @@ protected override ExpressionSyntax CreateCallSignature()
...
+                        {
+                            Length: 1
+                        } parameters
+                    } && parameters[0].Type is { SpecialType: SpecialType.System_Boolean }))
+                    {
+                        invocation = SyntaxFactory.InvocationExpression(
+                            SyntaxFactory.MemberAccessExpression(
+                                SyntaxKind.SimpleMemberAccessExpression,
+                                invocation,
+                                SyntaxFactory.IdentifierName("ConfigureAwait")),
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: nameof

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 4]
```diff
@@ -578,6 +578,29 @@ protected override ExpressionSyntax CreateCallSignature()
...
+                            SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(new[]
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: .SingletonSeparateList

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs [hunk 5]
```diff
@@ -71,15 +72,16 @@ internal abstract partial class CSharpSelectionResult : SelectionResult
...
         {
             var current = token.Parent;
             for (; current != null; current = current.Parent)
             {
                 if (current is MemberDeclarationSyntax ||
                     current is SimpleLambdaExpressionSyntax ||
                     current is ParenthesizedLambdaExpressionSyntax ||
-                    current is AnonymousMethodExpressionSyntax)
+                    current is AnonymousMethodExpressionSyntax ||
+                    current is LocalFunctionStatementSyntax)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: tests added for this?

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs [hunk 6]
```diff
@@ -95,6 +97,37 @@ protected override bool UnderAsyncAnonymousMethod(SyntaxToken token, SyntaxToken
                    current.GetLastToken().Span.End <= lastToken.Span.End;
         }
 
+        protected sealed override bool IsConfigureAwaitFalse(SyntaxNode node)
+        {
+            if (node is InvocationExpressionSyntax
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: i don't mind the huge pattern.  i just would prefer it not be in the if-condition.  can you just extract it and assign to a local?

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs [hunk 7]
```diff
@@ -95,6 +97,37 @@ protected override bool UnderAsyncAnonymousMethod(SyntaxToken token, SyntaxToken
...
+            {
+                Expression: MemberAccessExpressionSyntax
+                {
+                    Name:
+                    {
+                        Identifier:
+                        {
+                            ValueText: "ConfigureAwait"
+                        }
+                    }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: consider: `Name: { Identifier: { ValueText: nameof(ConfigureAwait) } }`.    This reads fine to me as an atomic description of hte name.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs [hunk 8]
```diff
@@ -95,6 +97,37 @@ protected override bool UnderAsyncAnonymousMethod(SyntaxToken token, SyntaxToken
...
+                },
+                ArgumentList:
+                {
+                    Arguments:
+                    {
+                        Count: 1
+                    } arguments
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: `ArgumentList: { Arguments: { Count: 1 } arguments }`

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs [hunk 9]
```diff
@@ -95,6 +97,37 @@ protected override bool UnderAsyncAnonymousMethod(SyntaxToken token, SyntaxToken
...
+            })
+            {
+                if (arguments[0].Expression is LiteralExpressionSyntax literalExpression
+                    && literalExpression.Kind() == SyntaxKind.FalseLiteralExpression)
+                    return true;
+            }
+
+            return false;
+        }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: this code is C# specific.  However, with SyntaxFacts, it feels like we could share this between C# and VB.

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs [hunk 10]
```diff
@@ -138,6 +139,29 @@ public bool ShouldPutAsyncModifier()
...
+
+            var span = TextSpan.FromBounds(firstToken.SpanStart, lastToken.Span.End);
+
+            foreach (var node in SemanticDocument.Root.DescendantNodesAndSelf())
+            {
+                if (!node.Span.OverlapsWith(span))
+                    continue;
+
+                else if (IsConfigureAwaitFalse(node) && !UnderAnonymousOrLocalMethod(node.GetFirstToken(), firstToken, lastToken))
+                {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: don't mix/match braces in the same cosntruct.  If the `else if` has braces, the `if` should have braces.

**Comment by CyrusNajmabadi** [comment]
Patterns are ute.  But i'd prefer just basic testing code that shares code across C# and VB.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: I tried doing this, but Vs have me warnings that it wasn't formatted correctly.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs (hunk 5 above)
Comment (YairHalberstadt) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: Indirectly Yes. It's the first test added in the cshsrp tests.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs (hunk 7 above)
Comment (YairHalberstadt) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: Again VS only allows patterns to be all one line, or every brace a new line.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: i say this honestly :)   I would gate the usae of these patterns on us fixing that that behavior of the formatter :)   I don't think it would be that hard.  I think we basically only need to add the "SuppressOperation" on htese pattersn if they're on a single line.

Comment (YairHalberstadt) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: I'm not sure I have the courage to face the formatter :-)

If I do fix it, how long would it take to pipe through though? Till the release of VS, or does Roslyn get it sooner?

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: the latter

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: yeah, we shouldn't gate on this.  but can you file an issue on this at least?  

It would take a while to pipe through.  but, in the meantime, you can just have code that doesn't match teh formatter.  we had that for a while with switch-expressions.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: it just goes with hte territory with adopting new lang features before the tooling is totally up to date with it :)

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: > I'm not sure I have the courage to face the formatter :-)

In this case i actually think it would be simple.  I can try to walk you through it if you want!

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs (hunk 10 above)
Comment (YairHalberstadt) [reply src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: ooh. Thats horrible. I definitely didn't write that on purpose!

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs [hunk 11]
```diff
@@ -165,39 +165,32 @@ public bool ShouldCallConfigureAwaitFalse()
...
-
             if (arguments.Count != 1)
             {
                 return false;
             }
 
             var argument = arguments[0];
-
             var expression = syntaxFacts.GetExpressionOfArgument(argument);
-
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: 👍

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 12]
```diff
@@ -578,6 +578,24 @@ protected override ExpressionSyntax CreateCallSignature()
                     return invocation;
                 }
 
+                if (this.CSharpSelectionResult.ShouldCallConfigureAwaitFalse())
+                {
+                    if (this.AnalyzerResult.ReturnType.GetMembers().Any(x => x is IMethodSymbol
+                    {
+                        Name: nameof(Task.ConfigureAwait),
+                        Parameters: { Length: 1 } parameters
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: 👍 i much prefer this.  a good balance of horizontal+vertical.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs [hunk 13]
```diff
@@ -578,6 +578,24 @@ protected override ExpressionSyntax CreateCallSignature()
...
+                    } && parameters[0].Type is { SpecialType: SpecialType.System_Boolean }))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs]: this doesn't feel like it needs a pattern.  just do `.Type.SpecialType == Boolean`.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionResult.cs]: #resolved thanks

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicMethodExtractor.VisualBasicCodeGenerator.vb [hunk 14]
```diff
@@ -331,6 +331,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExtractMethod
...
                 If Me.VBSelectionResult.ShouldPutAsyncModifier() Then
+                    If Me.VBSelectionResult.ShouldCallConfigureAwaitFalse() Then
+                        If AnalyzerResult.ReturnType.GetMembers().Any(
+                        Function(x)
+                            Dim method = TryCast(x, IMethodSymbol)
+                            If method Is Nothing Then
+                                Return False
+                            End If
+
+                            If Not CaseInsensitiveComparison.Equals(method.Name, "ConfigureAwait") Then
```
Comment (sharwell) [inline src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicMethodExtractor.VisualBasicCodeGenerator.vb]: 💡 `NameOf`

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicMethodExtractor.VisualBasicCodeGenerator.vb [hunk 15]
```diff
@@ -331,6 +331,38 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExtractMethod
...
+
+                            Return method.Parameters(0).Type.SpecialType = SpecialType.System_Boolean
+                        End Function) Then
+
+                            invocation = SyntaxFactory.InvocationExpression(
+                            SyntaxFactory.MemberAccessExpression(
+                                SyntaxKind.SimpleMemberAccessExpression,
+                                invocation,
+                                SyntaxFactory.Token(SyntaxKind.DotToken),
+                                SyntaxFactory.IdentifierName("ConfigureAwait")),
```
Comment (sharwell) [inline src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicMethodExtractor.VisualBasicCodeGenerator.vb]: 💡 `NameOf`

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs [hunk 16]
```diff
@@ -138,6 +139,61 @@ public bool ShouldPutAsyncModifier()
...
+
+            var arguments = syntaxFacts.GetArgumentsOfInvocationExpression(node);
+            if (arguments.Count != 1)
+            {
+                return false;
+            }
+
+            var argument = arguments[0];
+            var expression = syntaxFacts.GetExpressionOfArgument(argument);
+            return syntaxFacts.IsFalseLiteralExpression(expression);
```
Comment (sharwell) [inline src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: 💡 Since this is a refactoring (which only runs at the explicit request of a user), consider using the semantic model to evaluate the argument as a _constant expression_ with a false value. This is optional because it's a rare edge case.

Comment (sharwell) [inline src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: 💡 Consider adding a test for a named argument:

```
await x.ConfigureAwait(continueOnCapturedContext: false)
```

I have seen this pattern in real code. It should already be handled by `GetExpressionOfArgument` but we should make sure.

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs [hunk 17]
```diff
@@ -138,6 +139,61 @@ public bool ShouldPutAsyncModifier()
...
+
+            var invokedExpression = syntaxFacts.GetExpressionOfInvocationExpression(node);
+            if (!syntaxFacts.IsSimpleMemberAccessExpression(invokedExpression))
+            {
+                return false;
+            }
+
+            var name = syntaxFacts.GetNameOfMemberAccessExpression(invokedExpression);
+            var identifier = syntaxFacts.GetIdentifierOfSimpleName(name);
+            if (!syntaxFacts.StringComparer.Equals(identifier.ValueText, nameof(Task.ConfigureAwait)))
```
Comment (sharwell) [inline src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: 💡 This should work for `ValueTask` as well, but it would be good to add a test for this case (where the original code awaits the result of a `ValueTask` and uses `ConfigureAwait(false)`).

File: src/Features/Core/Portable/ExtractMethod/SelectionResult.cs (hunk 16 above)
Comment (YairHalberstadt) [reply src/Features/Core/Portable/ExtractMethod/SelectionResult.cs]: I think the only case where this is likely to occur is when there's a global constant which can be set to configure ContinueOnCapturedContext or not.

However I'm not even sure doing this is the right thing in such a case, because you probably want to use the global constant rather than false.

Given that this is extremely rare and would add reasonable complexity to an otherwise simple method, I've left it for now.

**Comment by allisonchou** [comment]
Thank you, @YairHalberstadt! 🎉
